src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -67,16 +77,27 @@ impl HummockSnapshotManager {
             let query_ids = core_guard.epoch_to_query_ids.get_mut(&epoch);
             if let Some(query_ids) = query_ids {
                 query_ids.remove(query_id);
-                if query_ids.is_empty() && epoch == core_guard.last_pinned {
-                    core_guard.is_outdated = true;
+                // Updated the outdated epoch.
+                if query_ids.is_empty() && core_guard.last_pinned == epoch {
+                    let epoch = Self::pin_epoch_with_retry(
+                        self.meta_client.clone(),
+                        Default::default(),
+                        usize::MAX,
+                        || false,
+                    )
+                    .await
```

Comment:
The question is: This PR still not differ too much from previous: pin_snapshot still in the critical path of query.

Reply:
For now, pin_snapshot is not in the get_epoch, but it is in unpin_snapshot.
I prepare to a new PR to wrap the epoch in an object to implement RAII and then change the unpin_snapshot method to avoid pin_snapshot in it.

Reply:
How to avoid pin_snapshot RFC and how this new design preserve the correctness? 

I am little confused about the flow now. Can you describe it more clearly or just use a diagram? https://sequencediagram.org/

Reply:
![image](https://user-images.githubusercontent.com/43447882/176152456-d74f7903-a1cd-4c33-a394-8a9a0da40b8c.png)

Oh..It looks like we don't need a background thread

Reply:
Elaborated more on the first event drive: in what situation why we need this update (when query ids of last_pinned is empty)?

Reply:
```
    pub async fn unpin_snapshot(&self, epoch: u64, query_id: &QueryId) -> SchedulerResult<()> {
        let min_epoch = async {
            // Decrease the ref count of corresponding epoch. If all last pinned epoch is dropped,
            // mark as outdated.
            let mut core_guard = self.core.lock().await;
            let query_ids = core_guard.epoch_to_query_ids.get_mut(&epoch);
            if let Some(query_ids) = query_ids {
                query_ids.remove(query_id);
                // Updated the outdated epoch.
                if query_ids.is_empty() && core_guard.last_pinned == epoch {                                <==============Here
                    Self::try_update_last_pinned(&mut core_guard, self.meta_client.clone()).await;    <==============Here
                }
            }
```
In this case, we need an update.
I try to comment the update and it cause the previous bug. It works when I add update here.
So I guess:In this case, we unpin the last query in last_pin, and this last_pin is outdated and we need to pin_snapshot to update it.